pkg/certificateprovider/azurekeyvault/provider.go

```diff
@@ -0,0 +1,368 @@
+	}
+	logrus.Infof("unmarshaled objects yaml, objectsArray %v", objects.Array)
+
+	keyVaultCerts := []types.KeyVaultCertificate{}
+	for i, object := range objects.Array {
+		var keyVaultCert types.KeyVaultCertificate
+		err = yaml.Unmarshal([]byte(object), &keyVaultCert)
+		if err != nil {
+			return nil, fmt.Errorf("unmarshal failed for keyVaultCerts at index %d, error: %w", i, err)
+		}
+		// remove whitespace from all fields in keyVaultCert
+		formatKeyVaultCertificate(&keyVaultCert)
+		if err = validate(keyVaultCert); err != nil {
+			return nil, wrapObjectTypeError(err, keyVaultCert.CertificateName, keyVaultCert.CertificateVersion)
+		}
+
+		keyVaultCerts = append(keyVaultCerts, keyVaultCert)
+	}
+
+	logrus.Infof("unmarshaled key vault objects, keyVaultObjects %v , count %v", keyVaultCerts, len(keyVaultCerts))
+
+	if len(keyVaultCerts) == 0 {
+		return nil, nil
+	}
+
+	// 2. initialize keyvault client
+
+	vaultURL, err := getVaultURL(keyvaultName, azureCloudEnv.KeyVaultDNSSuffix)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get vault")
+	}
+	logrus.Infof("vaultName %v, vaultURL %v", keyvaultName, *vaultURL)
+
+	kvClient, err := initializeKvClient(ctx, azureCloudEnv.KeyVaultEndpoint, azureCloudEnv.ActiveDirectoryEndpoint, tenantID, workloadIdentityClientID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get keyvault client")
+	}
+
+	// 3. for each object , get content bytes
+	files := []types.CertificateFile{}
+	for _, keyVaultCert := range keyVaultCerts {
+		logrus.Infof("fetching object from key vault, certName %v,  keyvault %v", keyVaultCert.CertificateName, keyvaultName)
+
+		resolvedKvCerts, err := resolveCertificateVersions(ctx, kvClient, keyVaultCert, *vaultURL)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, resolvedKvCert := range resolvedKvCerts {
+			// fetch the object from Key Vault
+			result, err := getCertificate(ctx, kvClient, *vaultURL, resolvedKvCert)
+			if err != nil {
+				return nil, err
+			}
+
+			for idx := range result {
+				r := result[idx]
+				objectContent := []byte(r.content)
+
+				file := types.CertificateFile{
+					Path:    resolvedKvCert.GetFileName(),
+					Content: objectContent,
+					Version: r.version,
+				}
+
+				files = append(files, file)
+				logrus.Infof("added file %v to response file", file.Path)
+			}
+		}
+	}
+	return files, nil
+}
+
+// formatKeyVaultCertificate formats the fields in KeyVaultCertificate
+func formatKeyVaultCertificate(object *types.KeyVaultCertificate) {
+	if object == nil {
+		return
+	}
+	objectPtr := reflect.ValueOf(object)
+	objectValue := objectPtr.Elem()
+
+	for i := 0; i < objectValue.NumField(); i++ {
+		field := objectValue.Field(i)
+		if field.Type() != reflect.TypeOf("") {
+			continue
+		}
+		str := field.Interface().(string)
+		str = strings.TrimSpace(str)
+		field.SetString(str)
+	}
+}
+
+// ParseAzureEnvironment returns azure environment by name
+func ParseAzureEnvironment(cloudName string) (*azure.Environment, error) {
```

Comment:
if this is not used outside of the package, it doesn't need to be exported

Reply:
fixed. thanks